cpp/src/arrow/filesystem/filesystem.h

```diff
@@ -122,18 +122,30 @@ ARROW_EXPORT std::ostream& operator<<(std::ostream& os, const FileInfo&);
 
 /// \brief File selector for filesystem APIs
 struct ARROW_EXPORT FileSelector {
-  /// The directory in which to select files.
-  /// If the path exists but doesn't point to a directory, this should be an error.
+  /// The path or glob in which to select files.
   std::string base_dir;
+  // Selector type of file selector
+  SelectorType type;
   /// The behavior if `base_dir` isn't found in the filesystem.  If false,
   /// an error is returned.  If true, an empty selection is returned.
   bool allow_not_found;
-  /// Whether to recurse into subdirectories.
-  bool recursive;
   /// The maximum number of subdirectories to recurse into.
   int32_t max_recursion;
 
-  FileSelector() : allow_not_found(false), recursive(false), max_recursion(INT32_MAX) {}
+  FileSelector()
+      : type(SelectorType::Directory), allow_not_found(false), max_recursion(0) {}
+
+  FileSelector(std::string base_dir, SelectorType type)
+      : base_dir(base_dir), type(type), allow_not_found(false), max_recursion(0) {}
+
+  bool recursive() const { return (type == SelectorType::Directory && max_recursion); }
```

Comment:
Isn't a glob selector recursive if it has `**`?

Reply:
I have a few questions, right now, we're not using that convention, but we could use it. Right now we are working only with '*' , but we can change it to '\*\*'. If we use '**' convention, do we want users to be able to specify the `max_recursion` or we assume we always have the maximum recursion? 

If we keep only '*' and max_recursion, recursive function would be like this: 
```
bool recursive() const { return ((type == SelectorType::Directory || (type == SelectorType::Glob && base_dir.back() == '*')) 
                                 && max_recursion); }
```

Reply:
Also, if we have this glob: `/A/*/B/`, does it mean a glob directory? So should we have the same result for `/A/*/B/` and `/A/*/B/**`?

Reply:
I think not supporting `**` is fine.  Let's just make sure we document that somewhere.  I had a brief discussion on this topic a few months ago on the Substrait Slack and Jacques guidance was "start narrow and expand from there" and so the spec currently says:

```
        // Consumers should support the POSIX syntax.  The recursive
        // globstar (**) may not be supported.
````

So we still meet the Substrait requirements even if we do not support `**`.